pallets/auctions/src/lib.rs

```diff
@@ -0,0 +1,693 @@
+        #[transactional]
+        pub fn cancel_auction(origin: OriginFor<T>, nft_id: NFTId) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let current_block = frame_system::Pallet::<T>::block_number();
+
+            let auction = Auctions::<T>::get(nft_id).ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+            // ensure auction creator is the caller
+            ensure!(
+                auction.creator == who,
+                Error::<T>::OnlyAuctionCreatorCanCancel
+            );
+
+            // ensure auction has not started
+            ensure!(
+                auction.start_block > current_block,
+                Error::<T>::CannotCancelInProcessAuction
+            );
+
+            // List nft as not for sale
+            T::NFTHandler::set_listed_for_sale(nft_id, false)?;
+
+            // Remove auction from storage
+            Auctions::<T>::remove(nft_id);
+
+            // Emit auction canceled event
+            Self::deposit_event(Event::AuctionCancelled { nft_id });
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(T::WeightInfo::add_bid())]
+        #[transactional]
+        pub fn add_bid(
+            origin: OriginFor<T>,
+            nft_id: NFTId,
+            amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let current_block = frame_system::Pallet::<T>::block_number();
+
+            // add bid to storage
+            Auctions::<T>::try_mutate(nft_id, |maybe_auction| -> DispatchResult {
+                let auction = maybe_auction
+                    .as_mut()
+                    .ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+                // ensure the caller is not the owner of NFT
+                ensure!(
+                    auction.creator != who.clone(),
+                    Error::<T>::OwnerCannotCreateBid
+                );
+
+                // ensure the auction period has commenced
+                ensure!(
+                    auction.start_block < current_block,
+                    Error::<T>::AuctionNotStarted
+                );
+
+                // ensure the auction period has not ended
+                ensure!(auction.end_block >= current_block, Error::<T>::AuctionEnded);
+
+                // ensure the bid is larger than the current highest bid
+                if let Some(highest_bid) = auction.bidders.get_highest_bid() {
+                    ensure!(amount > highest_bid.1, Error::<T>::InvalidBidAmount);
+                    // reject if user already has a bid
+                    ensure!(
+                        auction.bidders.find_bid(who.clone()).is_none(),
+                        Error::<T>::UserBidAlreadyExists
+                    );
+                } else {
+                    // ensure the bid amount is greater than start price
+                    ensure!(auction.start_price < amount, Error::<T>::InvalidBidAmount);
```

Comment:
I am not happy with the name InvalidBlidAmount but I don't know a better name for it :(